rfcs/0001-extended-lookup-tables.md

```diff
@@ -0,0 +1,530 @@
+and check for consistency of the evaluations and openings when verifying the
+proof.
+
+The `SideLoadedTable` polynomials are included in the proof, but we do so 'as
+if' they had been part of the verification index (ie. without any
+proof-specific openings), so that they may be reused across multiple different
+proofs without modification.
+
+The `joint_combiner` should be sampled using randomness that depends on the
+individual `SideLoadedTable(i,j)`s, to ensure that the side-loaded values
+cannot be chosen with prior knowledge `joint_combiner`.
+
+*Note: it may be useful for the verifier to pass the `SideLoadedTable(i,j)`s as
+public inputs to the proof, so that the circuit may assert that the table it
+had side-loaded was signed or otherwise correlated with a cryptographic
+commitment that it had access to. The details of this are deliberately elided
+in this RFC, since the principle and implementation are both fairly
+straightforward.*
+
+### Position-independence of side-loaded tables
+
+We want to ensure that side-loaded tables are maximally re-usable across
+different proofs, even where the 'position' of the same side-loaded table in
+the combined table may vary between proofs. For example, we could consider a
+'folding' circuit `C` which takes 2 side-loaded tables `t_in` and `t_out`,
+where we prove
+```
+C(t_0, t_1)
+C(t_1, t_2)
+C(t_2, t_3)
+...
+```
+and the 'result' of the fold is the final `t_n`. In each pair of executions in
+the sequence, we find that `t_i` is used as both `t_in` and `t_out`, but we
+would like to expose the same table as the value for each.
+
+To achieve this, we use a permutation argument to 'place' the values from each
+side-loaded table at the correct position in the final table. Recall that every
+value
+```
+LookupTable - FixedAndRuntimeTable
+```
+is either 0 (where no side-loaded table contributes a value) or the value of
+some row in `SideLoadedCombined(i) * joint_combiner` from the table `i`.
+
+For the `0` differences, we calculate a permutation between them all and the
+single constant `0`, to ensure that `LookupTable = FixedAndRuntimeTable`. For
+all other values, we set-up a permutation based upon the side-loaded table
+relevant to each value.
+
+Thus, we build the permutation accumulator
+```
+LookupPermutation
+ * ((LookupTable - FixedAndRuntimeTable) * joint_combiner^(-1)
+     + gamma + beta * Sigma(7))
+ * (SideLoadedCombined(0) + gamma + beta * Sigma(8))
+ * (SideLoadedCombined(1) + gamma + beta * Sigma(9))
+ * (SideLoadedCombined(2) + gamma + beta * Sigma(10))
+ * (SideLoadedCombined(3) + gamma + beta * Sigma(11))
+ =
+   LookupPermutation[prev]
+   * ((LookupTable - FixedAndRuntimeTable) * joint_combiner^(-1)
+       + gamma + x * beta * shift[7])
+   * (SideLoadedCombined(0) + gamma + x * beta * shift[8])
+   * (SideLoadedCombined(1) + gamma + x * beta * shift[9])
+   * (SideLoadedCombined(2) + gamma + x * beta * shift[10])
+   * (SideLoadedCombined(3) + gamma + x * beta * shift[11])
+```
```

Comment:
maybe it'd be nicer to show the accumulator before the polynomialized version

* `acc_0 = i`
* `acc_i = acc_{i-1} * f /g `

where

* `f = ((LookupTable - FixedAndRuntimeTable) * joint_combiner^(-1)
       + gamma + x * beta * shift[7])
   * (SideLoadedCombined(0) + gamma + x * beta * shift[8])
   * (SideLoadedCombined(1) + gamma + x * beta * shift[9])
   * (SideLoadedCombined(2) + gamma + x * beta * shift[10])
   * (SideLoadedCombined(3) + gamma + x * beta * shift[11])`
* `g = ((LookupTable - FixedAndRuntimeTable) * joint_combiner^(-1)
     + gamma + beta * Sigma(7))
 * (SideLoadedCombined(0) + gamma + beta * Sigma(8))
 * (SideLoadedCombined(1) + gamma + beta * Sigma(9))
 * (SideLoadedCombined(2) + gamma + beta * Sigma(10))
 * (SideLoadedCombined(3) + gamma + beta * Sigma(11))`